assets/src/stories-editor/components/editor-carousel/reordererIndicator.js

```diff
@@ -0,0 +1,185 @@
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * WordPress dependencies
+ */
+import { compose } from '@wordpress/compose';
+import { Component } from '@wordpress/element';
+import { withDispatch, withSelect } from '@wordpress/data';
+import { __, sprintf } from '@wordpress/i18n';
+import { Button, Draggable, DropZone, Tooltip } from '@wordpress/components';
+
+/**
+ * Parses drag & drop events to ensure the event contains valid transfer data.
+ *
+ * @param {Object} event
+ * @return {Object} Parsed event data.
+ */
+const parseDropEvent = ( event ) => {
+	let result = {
+		srcRootClientId: null,
+		srcClientId: null,
+		srcIndex: null,
+		type: null,
+	};
+
+	if ( ! event.dataTransfer ) {
+		return result;
+	}
+
+	try {
+		result = Object.assign( result, JSON.parse( event.dataTransfer.getData( 'text' ) ) );
+	} catch ( err ) {
+		return result;
+	}
+
+	return result;
+};
+
+/**
+ * A single item within the list of pages to be reordered.
+ *
+ * Re-uses existing Draggable and DropZone provided by WordPress
+ * in order to not re-invent the wheel.
+ */
+class ReordererIndicator extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.state = {
+			isDragging: false,
+		};
+
+		this.onDrop = this.onDrop.bind( this );
+	}
+
+	getInsertIndex( position ) {
+		const { index } = this.props;
+
+		if ( index !== undefined ) {
+			return position.x === 'right' ? index + 1 : index;
+		}
+	}
+
+	onDrop( event, position ) {
+		const { page: { clientId }, movePageToPosition, index } = this.props;
+		const { srcClientId, srcIndex, type } = parseDropEvent( event );
+
+		const isBlockDropType = ( dropType ) => dropType === 'block';
+		const isSameBlock = ( src, dst ) => src === dst;
+
+		if ( ! isBlockDropType( type ) || isSameBlock( srcClientId, clientId ) ) {
+			return;
+		}
+
+		const positionIndex = this.getInsertIndex( position );
+		const insertIndex = srcIndex < index ? positionIndex - 1 : positionIndex;
+		movePageToPosition( srcClientId, insertIndex );
+	}
+
+	componentDidMount() {
+		this.props.setOrder();
```

Comment:
The whole `setOrder` logic here seems... confusing.

I think we should handle this differently in a more simplified way: 

* `movePageToPosition` should just save the current block order in the block editor store permanently, not using our own store at all.
* No need for `setOrder`

In the regular reorder mode we could also think about simplifying the logic. Maybe using our own data store is a bit unnecessary.

Reply:
@swissspidy I think, `resetOrder` is useful in the current scenario 👍

Reply:
That's not what I meant.

I don't think this component should have a side effect of calling `setOrder` or `resetOrder` inside `componentDidMount`. 

I would remove this, and then instead of `movePageToPosition`, just use `moveBlockToPosition` from `core/block-editor`.